### PR TITLE
fix: conversion to handle nanosecond precision timestamp

### PR DESCRIPTION
**Pull request by russellwheatley**
fix: https://github.com/firebase/firebase-functions-python/issues/95

To test, I hardcoded some timestamps with nanosecond precision into the code and ran from the emulator:
Screenshot 2023-06-20 at 12 10 05

Here are the successful logs from the emulator changing the nanoseconds hardcoded:
Screenshot 2023-06-20 at 12 11 00
Screenshot 2023-06-20 at 12 12 41

I've also written tests for converting to nanosecond precision, microsecond precision and for detecting what type of timestamp it is [here](https://github.com/firebase/firebase-functions-python/pull/109/files#diff-0dcbadaa6c70afb6fea041dec04bb89c3cc7fe201fb7c3185a669efa49769f28R46-R105)

### CODE REVIEW

File: src/firebase_functions/private/util.py [hunk 1]
```diff
@@ -308,3 +309,48 @@ def firebase_config() -> None | FirebaseConfig:
...
+    s_ns = s_ns[:6]
+
+    # Put the string back together
+    s_processed = f"{s_datetime}.{s_ns}Z"
+
+    # Now parse the date-time string
+    event_time = _dt.datetime.strptime(s_processed, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+    # strptime assumes local time, we know it's UTC and so:
+    event_time = event_time.replace(tzinfo=_dt.timezone.utc)
```
Comment (Salakar) [inline src/firebase_functions/private/util.py:333]:
```suggestion
  # Separate the date and time part from the nanoseconds.
  datetime_str, nanosecond_str = timestamp_string.replace("Z", "").split(".")

  # Parse the date and time part of the string.
  event_time = datetime.strptime(datetime_str, "%Y-%m-%dT%H:%M:%S")

  # Add the microseconds and timezone.
  event_time = event_time.replace(microsecond=int(nanosecond_str[:6]), tzinfo=timezone.utc)
```

we could probably simplify to something like this

File: src/firebase_functions/private/util.py [hunk 2]
```diff
@@ -308,3 +309,48 @@ def firebase_config() -> None | FirebaseConfig:
...
+
+    return event_time
+
+
+def is_nanoseconds_timestamp(time: str) -> bool:
```
Comment (Salakar) [inline src/firebase_functions/private/util.py:338]:
```suggestion
def is_precision_timestamp(time: str) -> bool:
```

Nit, prefer 'precision' timestamp name over explicit unit of measurement

File: src/firebase_functions/private/util.py [hunk 3]
```diff
@@ -308,3 +309,48 @@ def firebase_config() -> None | FirebaseConfig:
...
+    """Return a bool which indicates if the timestamp is in nanoseconds"""
+    # Split the string into date-time and fraction of second
+    _, s_fraction = time.split(".")
+
+    # Split the fraction from the timezone specifier ('Z' or 'z')
+    s_fraction, _ = s_fraction.split(
+        "Z") if "Z" in s_fraction else s_fraction.split("z")
+
+    # If the fraction is 9 digits long, it's a nanosecond timestamp
+    return len(s_fraction) == 9
```
Comment (Salakar) [inline src/firebase_functions/private/util.py:348]:
```suggestion
    return len(s_fraction) > 6
```

Nit, anything with more than 6 digits is a precision timestamp, though in the issue we only ever see 9 digits

**Review by Salakar — APPROVED** [review]

File: src/firebase_functions/private/util.py [hunk 4]
```diff
@@ -310,6 +311,41 @@ def firebase_config() -> None | FirebaseConfig:
...
+def is_precision_timestamp(time: str) -> bool:
+    """Return a bool which indicates if the timestamp is in nanoseconds"""
+    # Split the string into date-time and fraction of second
+    _, s_fraction = time.split(".")
+
+    # Split the fraction from the timezone specifier ('Z' or 'z')
+    s_fraction, _ = s_fraction.split(
+        "Z") if "Z" in s_fraction else s_fraction.split("z")
+
+    # If the fraction is 9 digits long, it's a nanosecond timestamp
```
Comment (exaby73) [inline src/firebase_functions/private/util.py:337]: Can we change the comment to match the code below. Something like `If the fraction is more than 6 digits long, it's a nanosecond timestamp`?

Comment (exaby73) [inline src/firebase_functions/private/util.py:337]: Can we change the comment to match the code below. Something like If the fraction is more than 6 digits long, it's a nanosecond timestamp?

File: src/firebase_functions/private/util.py [hunk 5]
```diff
@@ -310,6 +311,41 @@ def firebase_config() -> None | FirebaseConfig:
     return FirebaseConfig(storage_bucket=json_data.get("storageBucket"))
 
 
+def nanoseconds_timestamp_conversion(time: str) -> _dt.datetime:
+    """Converts a nanosecond timestamp and returns a datetime object of the current time in UTC"""
+
+    # Separate the date and time part from the nanoseconds.
+    datetime_str, nanosecond_str = time.replace("Z", "").split(".")
```
Comment (exaby73) [inline src/firebase_functions/private/util.py:318]: Are we not handling the lowercase `z` here, similar to how it's handled in `is_precision_timestamp`?

**Review by exaby73 — CHANGES_REQUESTED** [review]
Small nits, but otherwise LGTM

File: src/firebase_functions/private/util.py (hunk 5 above)
Comment (russellwheatley) [reply src/firebase_functions/private/util.py]: Found an easier way to do it 
```
datetime_str, nanosecond_str = time.replace("Z", "").replace("z", "").split(".")
```

**Review by exaby73 — APPROVED** [review]

**Pull request merged**